koku/api/report/queries.py

```diff
@@ -682,37 +690,92 @@ def _percent_delta(self, a, b):
+                rank_orders.append(getattr(F(self._delta), self.order_direction)())
+        else:
+            rank_annotations = {self.order_field: self.report_annotations[self.order_field]}
+            rank_orders.append(getattr(F(self.order_field), self.order_direction)())
+
+        if tag_column in gb[0]:
+            rank_orders.append(self.get_tag_order_by(gb[0]))
+
+        # this is a sub-query, but not really.
+        # in the future, this could be accomplished using CTEs.
+        rank_by_total = Window(expression=Rank(), order_by=rank_orders)
+        ranks = (
+            query.annotate(**self.annotations)
+            .values(*group_by_value)
+            .annotate(**rank_annotations)
+            .annotate(rank=rank_by_total)
+        )
+
+        rankings = []
+        for rank in ranks:
+            rankings.insert((rank.get("rank") - 1), str(rank.get(group_by_value[0])))
+
+        return self._ranked_list(data, rankings)
+
+    def _ranked_list(self, data_list, ranks=None):
         """Get list of ranked items less than top.
 
         Args:
             data_list (List(Dict)): List of ranked data points from the same bucket
+            ranks (List): list of ranks to use; overrides ranking that may present in data_list.
         Returns:
             List(Dict): List of data points meeting the rank criteria
 
         """
         rank_limited_data = OrderedDict()
         date_grouped_data = self.date_group_data(data_list)
-        if data_list:
-            self.max_rank = max(entry.get("rank") for entry in data_list)
+
+        if ranks:
+            self.max_rank = len(ranks)
+        elif data_list:
+            self.max_rank = max(entry.get("rank", 0) for entry in data_list)
+
         is_offset = "offset" in self.parameters.get("filter", {})
 
         for date in date_grouped_data:
-            ranked_list = self._perform_rank_summation(date_grouped_data[date], is_offset)
+            ranked_list = self._perform_rank_summation(date_grouped_data[date], is_offset, ranks)
             rank_limited_data[date] = ranked_list
 
         return self.unpack_date_grouped_data(rank_limited_data)
 
-    def _perform_rank_summation(self, entry, is_offset):  # noqa: C901
-        """Do the actual rank limiting for rank_list."""
+    def _perform_rank_summation(self, entry, is_offset=False, ranks=[]):  # noqa: C901
+        """Do the rank limiting for _ranked_list().
+
+        Args:
+            entry (dict)
+            is_offset (bool)
+            ranks (list)
+        """
         other = None
         ranked_list = []
         others_list = []
         other_sums = {column: 0 for column in self._mapper.sum_columns}
         for data in entry:
             if other is None:
                 other = copy.deepcopy(data)
-            rank = data.get("rank")
+
+            if ranks:
+                rank = ranks.index(str(data.get(self._get_group_by()[0])))
```

Comment:
```suggestion
                rank = ranks.index(str(data.get(self._get_group_by()[0]))) + 1
```

Reply:
We either want to start at 1 here or change the if statement below

`if rank > self._offset and rank <= self._limit + self._offset:`

to

`if rank >= self._offset and rank < self._limit + self._offset:`

Reply:
I've applied the if-statement change. Thank you for catching this. ^_^